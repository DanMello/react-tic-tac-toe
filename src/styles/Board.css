.boardContainer * {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.boardContainer {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  position: absolute;
}

.boardRow {
  display: flex;
  height: 33.3%;
}