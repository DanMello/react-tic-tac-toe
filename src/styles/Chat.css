.chatContainer {
  background: white;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
}

.anotherContainer {
  display: flex;
  flex-direction: column;
  width: 100%;
  z-index: 999;
  background: white;
  position: fixed;
}

.subContainer{
  height: 40px;
  display: flex;
  align-items: center;
}

.closeButton {
  font-weight: bold;
  padding: 10px;
  font-size: 20px;
  color: #03a9f4; 
  cursor: pointer;
}

.sendMessage {
  font-weight: bold;
  padding: 10px;
  font-size: 18px;
  color: #03a9f4;
  cursor: pointer;  
}

.chatInput {
  flex: 1;
  border: none;
  border-radius: 0;
  outline: none;
  box-shadow: none;
  font-size: 16px;
  padding: 5px;
}

.line {
  width: 95%;
  align-self: center;
  border-bottom: 2px solid #ccc;
}

.mainMessageContainer {
  margin-top: 65px;
}

.messageContainerLeft {
  display: flex;
  justify-content: flex-start;
}

.messageContainerRight {
  display: flex;
  justify-content: flex-end;
}

.chatLeft, .chatRight {
  display: inline-block;
  padding: 10px;
  font-size: 20px;
  max-width: 70%;
  margin-top: 2px;
  border-radius: 10px;
}

.chatLeft {
  background: #ccc;
  color: #444;
  margin-left: 15px;
}

.chatRight {
  background: #03a9f4;
  color: white;
  margin-right: 15px;
}