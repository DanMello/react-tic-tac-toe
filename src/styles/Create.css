.container {
  display: flex;
  flex-direction: column;
  width: 100%;
  background: white;
  padding: 15px;
  border-radius: 5px;
}

.optionsContainer {
  display: flex;
  flex-direction: column;
}

.label {
  color: #444;
  font-size: 20px;
  margin-top: 10px;
}

.marginTop {
  margin-top: 10px;
}

.button {
  color: white;
  background: #03a9f4;
  border: none;
  outline: none;
  padding: 10px;
  font-weight: bold;
  border-radius: 5px;
  font-size: 20px;
  margin-top: 10px;
  cursor: pointer;
}

.moveLabel {
  font-size: 20px;
  font-weight: bold;
  color: #444;
  margin-left: 5px;
}

.input {
  margin-top: 10px;
  font-size: 16px;
}