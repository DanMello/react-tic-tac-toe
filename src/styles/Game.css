.mainContainer * {
  box-sizing: border-box;
  font-family: sans-serif;
}

.Container {
  /* display: flex;
  box-sizing: border-box;
  align-items: center;
  font-family: sans-serif;
  border-radius: 5px;
  flex-direction: column;
  padding-left: 15px;
  padding-right: 15px;
  padding-bottom: 15px;
  background: #03a9f4;
  width: 100%;
  position: relative; */
}