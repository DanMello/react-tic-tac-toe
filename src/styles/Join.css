.container {
  width: 100%;
  box-sizing: border-box;
  align-items: center;
  font-family: sans-serif;
  display: flex;
  flex-direction: column;
  background: #03a9f4;
  border-radius: 5px;
}

.searchingContainer {
  width: 100%;
  left: 0;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  padding-bottom: 15px;
  background: #03a9f4;
  z-index: 999;
  transition: top 0.5s;
}

.subContainer {
  width: 100%;
  padding-left: 15px;
  padding-bottom: 15px;
  padding-right: 15px;
  border-radius: 5px;
}

.noGames {
  color: white;
  display: flex;
  justify-content: center;
  padding-top: 20px;
}

.gamesContainer, .noGames {
  background: #03a9f4;
}

.paddingContainer {
  transition: padding 0.5s;
}

.loader {
  position: fixed;
  left: 50%;
  margin-left: -12.5px;
}

.loaderTwoContainer {
  display: flex;
  justify-content: center;
  transition: margin 0.5s;
}