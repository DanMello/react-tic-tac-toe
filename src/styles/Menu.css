.mainContainer {
  width: 100%;
}

.subContainer {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  color: white;
}

.heading {
  margin-top: 15px;
  margin-bottom: 0px;
  color: white;
  text-align: center;
}

.gameHeading {
  margin-top: 15px;
  color: white;
  text-align: center;
}

.usernameContainer {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
  margin-top: 10px;
}

.atme {
  font-weight: bold;
  font-size: 20px;
  color:#ffeb3b;
}

.username {
  border: none;
  background: none;
  padding-top: 5px;
  padding-bottom: 5px;
  padding-left: 2px;
  padding-right: 0px;
  margin: 0;
  font-size: 20px;
  color: white;
}

.error {
  background: rgb(223, 18, 52);
  width: 100%;
  text-align: center;
  color: white;
  padding: 2px;
  font-family: sans-serif;
  border-top-right-radius: 5px;
  border-top-left-radius: 5px;
}

.usernameError {
  color: orangered;
}

.errorContainer {
  display: flex;
  justify-content: center;
}

.newGamesContainer {
  display: flex;
  justify-content: center;
  background: #2ECC40;
  width: 100%;
  color: white;
  padding: 2px;
  font-family: sans-serif;
  border-top-right-radius: 5px;
  border-top-left-radius: 5px;
}

.dismiss {
  margin-left: 4px;
  color: white;
  cursor: pointer;
  text-decoration: underline;
}