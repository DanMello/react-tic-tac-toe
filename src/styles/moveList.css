.select {
  flex: 1;
  padding: 5px;
  margin-left: 5px;
}