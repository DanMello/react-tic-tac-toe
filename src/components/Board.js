import React, { useReducer } from 'react';
import Square from './Square';
import styles from 'styles/Board.css';
import calculateWinner from '../helpers/calculateWinner';
import chunkArray from '../helpers/chunkArray';
import gameReducer from '../reducer/gameReducer';
import MoveList from './MoveList';
import useTimeMachine from '../hooks/timeMachine';

function Board() {
  const [state, dispatch] = useReducer(gameReducer, {
    squares: Array(9).fill(null),
    isXNext: true
  });
  const {history, jumpToSquares, resetHistory} = useTimeMachine(state.squares, dispatch);
  const {squares, isXNext} = state;
  const winner = calculateWinner(squares);
  const chunkedArray = chunkArray(squares, 3);
  let status;

  function handleClick(i) {
    dispatch({type: 'SELECT_SQUARE', index: i});
  };

  function renderSquare(i) {
    return (
      <Square
        key={i}
        value={squares[i]}
        onClick={() => handleClick(i)}
      />
    );
  };

  function resetSquares() {
    dispatch({type: 'RESET_SQUARE'});
    resetHistory();
  };

  if (winner) {
    status = 'Winner: ' + winner;
  } else if (!winner && squares.every(item => item !== null)) {
    status = 'Game was a tie'
  } else {
    status = 'Next player: ' + (isXNext ? 'X' : 'O');
  };

  return (
    <div className={styles.boardContainer}>
      <div id='status' className={styles.status}>{status}</div>
      {chunkedArray.map((array, i) => {
        return (
          <div key={i} className={styles.boardRow}>
            {array.map((_, index) => {
              const chunkIndex = i > 0 ? ((i * 3) + index) : index;
              return renderSquare(chunkIndex);
            })}
          </div>
        );
      })}
      <div id='resetButton' className={styles.resetButton} onClick={resetSquares}>Reset</div>
      <MoveList history={history} jumpToSquares={jumpToSquares}/>
    </div>
  );
};

export default Board;