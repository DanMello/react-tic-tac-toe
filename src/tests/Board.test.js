import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import Board from '../components/Board';
import renderer from 'react-test-renderer';
import styles from 'styles/Board.css';

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it('Renders without crashing', () => {
  ReactDOM.render(<Board />, container);
});

it('Renders correctly', () => {
  const tree = renderer.create(<Board />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render 5 divs', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const divs = container.querySelectorAll('div');
  expect(divs.length).toEqual(5);
});

it('Each row should have 3 columns.', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const boardRows = container.querySelectorAll('.' + styles.boardRow);
  boardRows.forEach(row => {
    expect(row.children.length).toEqual(3);
  });
});

it('Board updates with X and O when user clicks on it and renders proper status.', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const squares = container.querySelectorAll('button');
  const status = container.querySelector('#status');
  expect(squares[0].textContent).toBe('');
  act(() => {
    squares[0].dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(status.innerHTML).toBe('Next player: O');
  expect(squares[0].textContent).toBe('X');
  expect(squares[1].textContent).toBe('');
  act(() => {
    squares[1].dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(status.innerHTML).toBe('Next player: X');
  expect(squares[1].textContent).toBe('O');
});

it('Should still only render 5 divs if all columns are selected and game was a tie.', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const squares = container.querySelectorAll('button');
  const orderOfClicksByIndex = [0,1,2,4,3,5,7,6,8];
  orderOfClicksByIndex.forEach(index => {
    act(() => {
      squares[index].dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
  });
  const divs = container.querySelectorAll('div');
  expect(divs.length).toEqual(5);
});

it('Clicking should only change the column that was clicked, dont change the numbers on chunkIndex Variable.', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const squares = container.querySelectorAll('button');
  act(() => {
    squares[2].dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  squares.forEach((_, index) => {
    if (index !== 2) {
      expect(squares[index].textContent).toBe('');
    }
  })
});

it('Should display winner when someone wins with their symbol', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const squares = container.querySelectorAll('button');
  const orderOfClicksforXtoWin = [0,1,4,2,8];
  orderOfClicksforXtoWin.forEach(index => {
    act(() => {
      squares[index].dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
  });
  const status = container.querySelector('#status');
  expect(status.innerHTML).toBe('Winner: X');
});

it('If game was a tie status should be Game was a tie.', () => {
  act(() => {
    ReactDOM.render(<Board />, container);
  });
  const squares = container.querySelectorAll('button');
  const orderOfClicksByIndex = [0,1,2,4,3,5,7,6,8];
  const status = container.querySelector('#status');
  orderOfClicksByIndex.forEach(index => {
    act(() => {
      squares[index].dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
  });
  expect(status.innerHTML).toBe('Game was a tie');
});