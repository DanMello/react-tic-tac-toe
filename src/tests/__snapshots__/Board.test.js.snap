// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="boardContainer"
>
  <div
    className="status"
    id="status"
  >
    Next player: X
  </div>
  <div
    className="boardRow"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="boardRow"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="boardRow"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="resetButton"
    id="resetButton"
    onClick={[Function]}
  >
    Reset
  </div>
  <div>
    <li>
      <button
        onClick={[Function]}
      >
        Go to game start
      </button>
    </li>
  </div>
</div>
`;
